### PR TITLE
feat: reverse geolocation per polygon strategy

### LINKED ISSUE #1301: Optimize GBFS geolocation extraction: Use OSM polygons as query cache
### Describe the problem

Geolocation extraction is a great feature, but it consumes a high amount of resources. In this issue, we will optimize the extraction of locations on the GBFS feeds. As it has not been proven that the proposed solution will increase performance until it is implemented and tested, as part of this issue, we are adding the ability to "configure" the extraction strategy. This will keep the code intact, and we will be able to conduct live tests to gather more accurate data for comparison.

### Proposed solution

Currently, we query the database once per geographic point. For GBFS, many geographic points are "clustered". This is primarily true in the case of vehicle positions as they tend to be clustered across regions in the same city. In many cities, the scooters/bikes/etc are only allowed to be parked in certain areas, which increases the points per cluster. This is empirical information.

Tasks:
- Adjust the code to allow multiple location extraction strategies by using a cloud function parameter or a pub/sub event parameter. Consider adding a default in case no strategy is passed.
- Having the list of points to extract the geolocation, implement the following strategy: 
  1. Query the DB for one point  
  2. Get the list of OSM IDs
  3. From the OSM IDs, select the one with the highest level and retrieve the polygon associated 
  4. Iterate over the list and remove from the list all the points that are in the retrieved polygon
  5. Go to step 1 until the list is completed
- Test the strategy with a few feeds and adjust it if needed. Consider twinking the comparison of the polygon by comparing all OSM polygons
- Consider adjusting the CPU and memory function if needed, as the local processing will demand more resources.

### Alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by davidgamez**
**Summary:**

Closes #1301, #1289, #965
**Significant changes in this PR:**
- Introduced `per_polygon` strategy
   - This strategy, described [here](), looks for clustering points within a polygon to reduce the amount of DB calls. This works for cases where the points are close together and the geolocation query has "high" admin levels in their respective country. In cases where the admin levels are below the municipality, the algorithm performs as a "per_point" strategy.
-  Fixed duplicate key value violates unique constraint. This was happening because the per-point strategy was attempting to create the same location multiple times within the same database session.
- Added a guard against multiple executions of the function within the same day.
- Stop using stop caching for GBFS, as vehicle positions change constantly.
- Fixed duplicate admin levels. This issue had two causes(#965):
  - The difference between the two representations (plane 2D and geospatial 3D) was causing duplicate intersections between polygons on large areas and edges. To fix this, the code now is casting the geometry as Geography. 
  - OSM doesn’t enforce uniqueness at each level; different contributors can model administrative boundaries differently. To fix this, a tie-break mechanism was implemented.

**Minor fixes**
- Fixed refresh materialized view function name in cloud task
- Missing GCP function region and project ID parameters

**Out of scope:**
- Address timeout for large feeds. This should be part of #1324 
- Create a configuration service. This is part of #1325

**Expected behavior:** 

- ✅  A reduced running time for feeds with a high number of stops per polygon.
- ✅  Reverse geolocation function runs a maximum of once per feed day and feed ID.

## Performance results

The performance results were collected for a few feeds without using the geometry cache.

| Feed |    Stops | Locations |    Strategy    | Time | Memory(peak) | Notes |
| -------- | ------- | ------- |------- | ------- |------- |------- |
| **GTFS** |  ---   |  ---      | --- | --- | --- |---|
| mdb-2014 |  308611   |  225    |Per-Point | 33130s(9.2h) | 129.57 MB ||
|  |     |        | Per-Polygon | 25870s(7.2h) | 133.11 MB ||
| --- |  ---   |  ---      | --- | --- | --- |---|
| mdb-1139 |   663  |  324   |Per-Point | 327s(5.45m) | 19.48 MB ||
|  |     |     | Per-Polygon | 100(1.6m) | 129.57 MB ||
| --- |  ---   |  ---      | --- | --- | --- |---|
| mdb-2825 |   10985  |  1541   |Per-Point | 1052.04(17.5)  | 25.55  MB ||
|  |     |     | Per-Polygon | 348.9s(5.8m) | 34.48 MB ||
| **GBFS** |  ---   |  ---      | --- | --- | --- |--- |
| gbfs-flamingo_auckland |   1539  |  1   |Per-Point | 37.2s | 1.34 MB ||
|  |     |     | Per-Polygon | 37.9 | 1.24 MB |All points reverse geolocation queries returns adminlevel 4|
| gbfs-hellocycling |   11777  |  241   |Per-Point | 337.76s(5.6m) | 14.16 MB ||
|  |     |     | Per-Polygon | 31.59 | 9.81 MB ||

**Testing tips:**
Local testing
- Use reverse_geolocation_process_verifier.py for local testing

## From our AI friend:
This pull request introduces new geospatial utilities and refactors location handling logic to improve reverse geolocation and location grouping features. It adds helper functions for working with geopolygons, enhances enum conversion error reporting, and expands test coverage for location utilities. Several dependencies and configuration files are updated to support these changes.

### Geospatial utilities and location logic enhancements

* Added new helper functions in `locations.py` for converting geometry types (`to_shapely`), selecting geopolygons by admin level (`select_highest_level_polygon`, `select_lowest_level_polygon`), extracting country codes from polygons, and querying geopolygons that cover a point with BigQuery-compatible semantics.
* Introduced new reverse geolocation utilities in `location_group_utils.py`, including functions for extracting location aggregates, creating/updating stop groups, and managing location group entities. These functions leverage the new geopolygon helpers for more robust location assignment and grouping. [[1]](diffhunk://#diff-15a693536ddc1d15cf2a85e55f97cedaf00cdf8edbae48c1d4380a2462072f01R1-R18) [[2]](diffhunk://#diff-15a693536ddc1d15cf2a85e55f97cedaf00cdf8edbae48c1d4380a2462072f01R145-R295)

### Test coverage improvements

* Expanded `test_locations.py` with comprehensive tests for the new geospatial helper functions, including edge cases for admin level selection and country code extraction from geopolygons.

### Dependency and configuration updates

* Added `shapely` to `requirements.txt` to support geometry conversions and spatial operations.
* Updated `.coveragerc` to omit coverage for files in `*/shared/common/*`, reflecting the new shared helpers.

### Error handling and logging improvements

* Improved error reporting in enum conversion by logging failed conversions in `transform.py`.
* Updated logger initialization to import `google.cloud.logging` only when needed, improving startup reliability.

### Minor fixes

* Fixed a typo in the Cloud Function URL construction in `gcp_utils.py` (`tasks_executor` instead of `tasks-executor`).
* Added missing import for logging in `transform.py`.
* Cleaned up unused imports in `logger.py`.
* Added new enum values and documentation for reverse geocoding strategies in `locations.py`.

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `./scripts/api-tests.sh` to make sure you didn't break anything
- [x] Add or update any needed documentation to the repo
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: functions-python/reverse_geolocation/src/strategy_extraction_per_polygon.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+) -> None:
+    """
+    Batch points by their containing geopolygon and compute one location aggregate per group.
+    """
+    remaining_stops_df = stops_df.copy()
+    processed_groups: set[str] = set()
+
+    total_stop_count = len(remaining_stops_df)
+    last_seen_count = total_stop_count
+    batch_size = total_stop_count / 20  # Process 5% of the total stops in each batch
```
Comment (jcpitre) [inline functions-python/reverse_geolocation/src/strategy_extraction_per_polygon.py:62]: Does this work for 19 stops? I saw that you did this in extract_location_aggregates_per_point:
    batch_size = max(
        int(total_stop_count / 20), total_stop_count

**Review by qcdyx — APPROVED** [review]
Excellent work!

File: functions-python/helpers/locations.py [hunk 2]
```diff
@@ -133,3 +144,90 @@ def translate_feed_locations(feed: Feed, location_translations: Dict):
...
+        return None
+
+    # Prefer the one with the lowest admin_level (most local)
+    lowest_admin_level_polygon = select_lowest_level_polygon(country_polygons)
+    return lowest_admin_level_polygon.iso_3166_1_code
+
+
+def get_geopolygons_covers(stop_point: WKTElement, db_session: Session):
+    """
+    Get all geopolygons that cover a given point using BigQuery-compatible semantics.
```
Comment (cka-y) [inline functions-python/helpers/locations.py:218]: [question] what does this have to do w big query?

File: functions-python/reverse_geolocation/src/location_group_utils.py [hunk 3]
```diff
@@ -130,3 +145,208 @@ def geopolygons_as_string(geopolygons: List[Geopolygon]) -> str:
...
+
+
+def geodesic_area_m2(geom) -> float:
+    """Return absolute geodesic area in m² for a Shapely geometry (Polygon/MultiPolygon)."""
+    # pyproj can handle GeoJSON-like mappings, including MultiPolygons
+    area, _perim = GEOD.geometry_area_perimeter(mapping(geom))
+    return abs(area)
+
+
+def resolve_same_level_conflicts(items: List[Geopolygon]) -> Geopolygon:
```
Comment (cka-y) [inline functions-python/reverse_geolocation/src/location_group_utils.py:178]: great addition 🤩

**Review by cka-y — APPROVED** [review]
LGTM 🏆

File: functions-python/helpers/locations.py (hunk 2 above)
Comment (davidgamez) [reply functions-python/helpers/locations.py:218]: The fact that we have to _ cast(Geopolygon.geometry, Geography(srid=4326))_ as from the values in BigQuery tables are Geography instead of geometry
